File: typescript/provenance-mark/src/validate.ts
/**
 * Validation support for provenance mark chains.
 */

import { CborDate } from '@bc/dcbor';
import { ProvenanceMark } from './provenance-mark.js';
import { ProvenanceMarkError } from './error.js';
import { bytesToHex, compareBytes } from './utils.js';

/**
 * Output formats for validation reports.
 */
export enum ValidationReportFormat {
  /** Human-readable text format. */
  Text = 'Text',
  /** Compact JSON format with no whitespace. */
  JsonCompact = 'JsonCompact',
  /** Pretty-printed JSON format with indentation. */
  JsonPretty = 'JsonPretty',
}

export type ValidationIssue =
  /** Hash mismatch between consecutive marks. */
  | { type: 'HashMismatch'; expected: Uint8Array; actual: Uint8Array }
  /** Key mismatch between consecutive marks. */
  | { type: 'KeyMismatch' }
  /** Sequence number gap. */
  | { type: 'SequenceGap'; expected: number; actual: number }
  /** Date ordering violation. */
  | { type: 'DateOrdering'; previous: CborDate; next: CborDate }
  /** Non-genesis mark at sequence 0. */
  | { type: 'NonGenesisAtZero' }
  /** Invalid genesis key. */
  | { type: 'InvalidGenesisKey' };

export function validationIssueToString(issue: ValidationIssue): string {
  switch (issue.type) {
    case 'HashMismatch':
      return `hash mismatch: expected ${bytesToHex(issue.expected)}, got ${bytesToHex(issue.actual)}`;
    case 'KeyMismatch':
      return 'key mismatch: current hash was not generated from next key';
    case 'SequenceGap':
      return `sequence number gap: expected ${issue.expected}, got ${issue.actual}`;
    case 'DateOrdering':
      return `date must be equal or later: previous is ${issue.previous.toString()}, next is ${issue.next.toString()}`;
    case 'NonGenesisAtZero':
      return 'non-genesis mark at sequence 0';
    case 'InvalidGenesisKey':
      return 'genesis mark must have key equal to chain_id';
  }
}

function validationIssueToJSON(issue: ValidationIssue): unknown {
  switch (issue.type) {
    case 'HashMismatch':
      return {
        type: 'HashMismatch',
        data: {
          expected: bytesToHex(issue.expected),
          actual: bytesToHex(issue.actual),
        },
      };
    case 'KeyMismatch':
      return { type: 'KeyMismatch' };
    case 'SequenceGap':
      return {
        type: 'SequenceGap',
        data: {
          expected: issue.expected,
          actual: issue.actual,
        },
      };
    case 'DateOrdering':
      return {
        type: 'DateOrdering',
        data: {
          previous: issue.previous.toString(),
          next: issue.next.toString(),
        },
      };
    case 'NonGenesisAtZero':
      return { type: 'NonGenesisAtZero' };
    case 'InvalidGenesisKey':
      return { type: 'InvalidGenesisKey' };
  }
}

/**
 * A provenance mark plus the issues flagged during validation.
 */
export class FlaggedMark {
  readonly #mark: ProvenanceMark;
  readonly #issues: ValidationIssue[];

  private constructor(mark: ProvenanceMark, issues: ValidationIssue[]) {
    this.#mark = mark;
    this.#issues = issues;
  }

  static create(mark: ProvenanceMark): FlaggedMark {
    return new FlaggedMark(mark, []);
  }

  static withIssue(mark: ProvenanceMark, issue: ValidationIssue): FlaggedMark {
    return new FlaggedMark(mark, [issue]);
  }

  get mark(): ProvenanceMark {
    return this.#mark;
  }

  get issues(): readonly ValidationIssue[] {
    return [...this.#issues];
  }

  toJSON(): unknown {
    return {
      mark: this.#mark.toUrString(),
      issues: this.#issues.map(validationIssueToJSON),
    };
  }
}

/**
 * Report for a contiguous sequence of marks within a chain.
 */
export class SequenceReport {
  readonly #startSeq: number;
  readonly #endSeq: number;
  readonly #marks: FlaggedMark[];

  constructor(startSeq: number, endSeq: number, marks: FlaggedMark[]) {
    this.#startSeq = startSeq;
    this.#endSeq = endSeq;
    this.#marks = marks;
  }

  get startSeq(): number {
    return this.#startSeq;
  }

  get endSeq(): number {
    return this.#endSeq;
  }

  get marks(): readonly FlaggedMark[] {
    return [...this.#marks];
  }

  toJSON(): unknown {
    return {
      start_seq: this.#startSeq,
      end_seq: this.#endSeq,
      marks: this.#marks.map((m) => m.toJSON()),
    };
  }
}

/**
 * Report for all marks that share the same chain ID.
 */
export class ChainReport {
  readonly #chainId: Uint8Array;
  readonly #hasGenesis: boolean;
  readonly #marks: ProvenanceMark[];
  readonly #sequences: SequenceReport[];

  constructor(
    chainId: Uint8Array,
    hasGenesis: boolean,
    marks: ProvenanceMark[],
    sequences: SequenceReport[],
  ) {
    this.#chainId = chainId;
    this.#hasGenesis = hasGenesis;
    this.#marks = marks;
    this.#sequences = sequences;
  }

  get chainId(): Uint8Array {
    return new Uint8Array(this.#chainId);
  }

  get hasGenesis(): boolean {
    return this.#hasGenesis;
  }

  get marks(): readonly ProvenanceMark[] {
    return [...this.#marks];
  }

  get sequences(): readonly SequenceReport[] {
    return [...this.#sequences];
  }

  /**
   * Return the chain ID as a hexadecimal string.
   */
  chainIdHex(): string {
    return bytesToHex(this.#chainId);
  }

  toJSON(): unknown {
    return {
      chain_id: bytesToHex(this.#chainId),
      has_genesis: this.#hasGenesis,
      marks: this.#marks.map((m) => m.toUrString()),
      sequences: this.#sequences.map((s) => s.toJSON()),
    };
  }
}

/**
 * Complete validation report for a set of provenance marks.
 */
export class ValidationReport {
  readonly #marks: ProvenanceMark[];
  readonly #chains: ChainReport[];

  private constructor(marks: ProvenanceMark[], chains: ChainReport[]) {
    this.#marks = marks;
    this.#chains = chains;
  }

  get marks(): readonly ProvenanceMark[] {
    return [...this.#marks];
  }

  get chains(): readonly ChainReport[] {
    return [...this.#chains];
  }

  // -----------------------------------------------------------------------
  // Validation
  // -----------------------------------------------------------------------

  /**
   * Validate a collection of provenance marks.
   */
  static validate(marks: ProvenanceMark[]): ValidationReport {
    // Deduplicate exact duplicates by message equality
    const seen = new Set<string>();
    const deduplicatedMarks: ProvenanceMark[] = [];
    for (const mark of marks) {
      const msg = bytesToHex(mark.message());
      if (!seen.has(msg)) {
        seen.add(msg);
        deduplicatedMarks.push(mark);
      }
    }

    // Bin marks by chain ID
    const chainBins = new Map<string, ProvenanceMark[]>();
    for (const mark of deduplicatedMarks) {
      const key = bytesToHex(mark.chainId);
      let bin = chainBins.get(key);
      if (bin === undefined) {
        bin = [];
        chainBins.set(key, bin);
      }
      bin.push(mark);
    }

    // Process each chain
    const chains: ChainReport[] = [];
    for (const [, chainMarks] of chainBins) {
      // Sort by sequence number
      chainMarks.sort((a, b) => a.seq - b.seq);

      // Check for genesis mark
      const first = chainMarks[0];
      const hasGenesis =
        first !== undefined && first.seq === 0 && first.isGenesis();

      // Build sequence bins
      const sequences = ValidationReport.buildSequenceBins(chainMarks);

      chains.push(
        new ChainReport(
          first!.chainId,
          hasGenesis,
          [...chainMarks],
          sequences,
        ),
      );
    }

    // Sort chains by chain ID bytes for consistent output
    chains.sort((a, b) => compareBytes(a.chainId, b.chainId));

    return new ValidationReport(deduplicatedMarks, chains);
  }

  private static buildSequenceBins(
    marks: ProvenanceMark[],
  ): SequenceReport[] {
    const sequences: SequenceReport[] = [];
    let currentSequence: FlaggedMark[] = [];

    for (let i = 0; i < marks.length; i++) {
      const mark = marks[i]!;

      if (i === 0) {
        // First mark starts a sequence
        currentSequence.push(FlaggedMark.create(mark));
      } else {
        const prev = marks[i - 1]!;

        // Check if this mark follows the previous one
        try {
          prev.assertPrecedes(mark);
          // Continues the current sequence
          currentSequence.push(FlaggedMark.create(mark));
        } catch (e: unknown) {
          // Breaks the sequence - save current and start new
          if (currentSequence.length > 0) {
            sequences.push(
              ValidationReport.createSequenceReport(currentSequence),
            );
          }

          // Extract ValidationIssue from the error
          let issue: ValidationIssue;
          if (
            e instanceof ProvenanceMarkError &&
            e.code === 'Validation' &&
            e.validationIssue !== undefined
          ) {
            issue = e.validationIssue as ValidationIssue;
          } else {
            issue = { type: 'KeyMismatch' };
          }

          // Start new sequence with this mark, flagged with the issue
          currentSequence = [FlaggedMark.withIssue(mark, issue)];
        }
      }
    }

    // Add the final sequence
    if (currentSequence.length > 0) {
      sequences.push(
        ValidationReport.createSequenceReport(currentSequence),
      );
    }

    return sequences;
  }

  private static createSequenceReport(
    marks: FlaggedMark[],
  ): SequenceReport {
    const startSeq = marks[0]?.mark.seq ?? 0;
    const endSeq = marks[marks.length - 1]?.mark.seq ?? 0;
    return new SequenceReport(startSeq, endSeq, marks);
  }

  // -----------------------------------------------------------------------
  // Formatting
  // -----------------------------------------------------------------------

  /**
   * Format the validation report as text or JSON.
   */
  format(fmt: ValidationReportFormat): string {
    switch (fmt) {
      case ValidationReportFormat.Text:
        return this.formatText();
      case ValidationReportFormat.JsonCompact:
        return JSON.stringify(this.toJSON());
      case ValidationReportFormat.JsonPretty:
        return JSON.stringify(this.toJSON(), null, 2);
    }
  }

  private formatText(): string {
    if (!this.isInteresting()) {
      return '';
    }

    const lines: string[] = [];

    // Report summary
    lines.push(`Total marks: ${this.#marks.length}`);
    lines.push(`Chains: ${this.#chains.length}`);
    lines.push('');

    // Report each chain
    for (let chainIdx = 0; chainIdx < this.#chains.length; chainIdx++) {
      const chain = this.#chains[chainIdx]!;

      // Show short chain ID (first 4 bytes = 8 hex chars)
      const chainIdHex = chain.chainIdHex();
      const shortChainId =
        chainIdHex.length > 8 ? chainIdHex.slice(0, 8) : chainIdHex;

      lines.push(`Chain ${chainIdx + 1}: ${shortChainId}`);

      if (!chain.hasGenesis) {
        lines.push('  Warning: No genesis mark found');
      }

      // Report each sequence
      for (const seq of chain.sequences) {
        // Report each mark in the sequence
        for (const flaggedMark of seq.marks) {
          const mark = flaggedMark.mark;
          const shortId = mark.identifier();
          const seqNum = mark.seq;

          // Build the mark line with annotations
          const annotations: string[] = [];

          // Check if it's genesis
          if (mark.isGenesis()) {
            annotations.push('genesis mark');
          }

          // Add issue annotations
          for (const issue of flaggedMark.issues) {
            let issueStr: string;
            switch (issue.type) {
              case 'SequenceGap':
                issueStr = `gap: ${issue.expected} missing`;
                break;
              case 'DateOrdering':
                issueStr = `date ${issue.previous.toString()} < ${issue.next.toString()}`;
                break;
              case 'HashMismatch':
                issueStr = 'hash mismatch';
                break;
              case 'KeyMismatch':
                issueStr = 'key mismatch';
                break;
              case 'NonGenesisAtZero':
                issueStr = 'non-genesis at seq 0';
                break;
              case 'InvalidGenesisKey':
                issueStr = 'invalid genesis key';
                break;
            }
            annotations.push(issueStr);
          }

          // Format the line
          if (annotations.length === 0) {
            lines.push(`  ${seqNum}: ${shortId}`);
          } else {
            lines.push(
              `  ${seqNum}: ${shortId} (${annotations.join(', ')})`,
            );
          }
        }
      }

      lines.push('');
    }

    return lines.join('\n').trimEnd();
  }

  /**
   * Check if the validation report contains interesting information.
   * Returns false for a single perfect chain with no issues, true otherwise.
   */
  private isInteresting(): boolean {
    // Not interesting if empty
    if (this.#chains.length === 0) {
      return false;
    }

    // Check if any chain is missing genesis
    for (const chain of this.#chains) {
      if (!chain.hasGenesis) {
        return true;
      }
    }

    // Not interesting if single chain with single perfect sequence
    if (this.#chains.length === 1) {
      const chain = this.#chains[0]!;
      if (chain.sequences.length === 1) {
        const seq = chain.sequences[0]!;
        // Check if the sequence has no issues
        if (seq.marks.every((m) => m.issues.length === 0)) {
          return false;
        }
      }
    }

    return true;
  }

  /**
   * Check if the validation report has any issues.
   * Returns true if there are validation issues, missing genesis,
   * multiple chains, or multiple sequences.
   */
  hasIssues(): boolean {
    // Missing genesis is considered an issue
    for (const chain of this.#chains) {
      if (!chain.hasGenesis) {
        return true;
      }
    }

    // Check for validation issues in marks
    for (const chain of this.#chains) {
      for (const seq of chain.sequences) {
        for (const mark of seq.marks) {
          if (mark.issues.length > 0) {
            return true;
          }
        }
      }
    }

    // Multiple chains or sequences are also considered issues
    if (this.#chains.length > 1) {
      return true;
    }

    if (
      this.#chains.length === 1 &&
      this.#chains[0]!.sequences.length > 1
    ) {
      return true;
    }

    return false;
  }

  toJSON(): unknown {
    return {
      marks: this.#marks.map((m) => m.toUrString()),
      chains: this.#chains.map((c) => c.toJSON()),
    };
  }
}

File: python/provenance-mark/src/provenance_mark/_validate.py
"""Validation report model for provenance-mark."""

from __future__ import annotations

import json
from dataclasses import dataclass
from enum import Enum

from typing import TYPE_CHECKING

from dcbor import Date

from ._error import Error

if TYPE_CHECKING:
    from ._mark import ProvenanceMark


class ValidationReportFormat(Enum):
    """Output format for validation reports."""

    Text = "Text"
    JsonCompact = "JsonCompact"
    JsonPretty = "JsonPretty"


@dataclass(frozen=True, slots=True)
class ValidationIssue:
    """One validation issue associated with a mark."""

    issue_type: str
    data: dict[str, object] | None = None

    def to_json_obj(self) -> dict[str, object]:
        result: dict[str, object] = {"type": self.issue_type}
        if self.data is not None:
            payload: dict[str, object] = {}
            for key, value in self.data.items():
                if isinstance(value, bytes):
                    payload[key] = value.hex()
                else:
                    payload[key] = str(value) if hasattr(value, "datetime_value") else value
            result["data"] = payload
        return result

    def __str__(self) -> str:
        if self.issue_type == "HashMismatch":
            assert self.data is not None
            return (
                "hash mismatch: expected "
                f"{bytes(self.data['expected']).hex()}, got "
                f"{bytes(self.data['actual']).hex()}"
            )
        if self.issue_type == "KeyMismatch":
            return "key mismatch: current hash was not generated from next key"
        if self.issue_type == "SequenceGap":
            assert self.data is not None
            return (
                f"sequence number gap: expected {self.data['expected']}, "
                f"got {self.data['actual']}"
            )
        if self.issue_type == "DateOrdering":
            assert self.data is not None
            return (
                "date must be equal or later: previous is "
                f"{self.data['previous']}, next is {self.data['next']}"
            )
        if self.issue_type == "NonGenesisAtZero":
            return "non-genesis mark at sequence 0"
        if self.issue_type == "InvalidGenesisKey":
            return "genesis mark must have key equal to chain_id"
        return self.issue_type


def validation_issue_to_string(issue: ValidationIssue) -> str:
    return str(issue)


@dataclass(frozen=True, slots=True)
class FlaggedMark:
    """A mark plus its validation issues."""

    mark: ProvenanceMark
    issues: tuple[ValidationIssue, ...] = ()

    @staticmethod
    def new(mark: ProvenanceMark) -> FlaggedMark:
        return FlaggedMark(mark)

    @staticmethod
    def with_issue(mark: ProvenanceMark, issue: ValidationIssue) -> FlaggedMark:
        return FlaggedMark(mark, (issue,))

    def to_json_obj(self) -> dict[str, object]:
        return {
            "mark": self.mark.ur_string(),
            "issues": [issue.to_json_obj() for issue in self.issues],
        }


@dataclass(frozen=True, slots=True)
class SequenceReport:
    """A contiguous sequence of marks within one chain."""

    start_seq: int
    end_seq: int
    marks: tuple[FlaggedMark, ...]

    def to_json_obj(self) -> dict[str, object]:
        return {
            "start_seq": self.start_seq,
            "end_seq": self.end_seq,
            "marks": [mark.to_json_obj() for mark in self.marks],
        }


@dataclass(frozen=True, slots=True)
class ChainReport:
    """Validation report for one chain ID."""

    chain_id: bytes
    has_genesis: bool
    marks: tuple[ProvenanceMark, ...]
    sequences: tuple[SequenceReport, ...]

    def chain_id_hex(self) -> str:
        return self.chain_id.hex()

    def to_json_obj(self) -> dict[str, object]:
        return {
            "chain_id": self.chain_id.hex(),
            "has_genesis": self.has_genesis,
            "marks": [mark.ur_string() for mark in self.marks],
            "sequences": [sequence.to_json_obj() for sequence in self.sequences],
        }


@dataclass(frozen=True, slots=True)
class ValidationReport:
    """Full validation report for a set of provenance marks."""

    marks: tuple[ProvenanceMark, ...]
    chains: tuple[ChainReport, ...]

    @staticmethod
    def validate(marks: list[ProvenanceMark]) -> ValidationReport:
        seen: set[ProvenanceMark] = set()
        deduplicated: list[ProvenanceMark] = []
        for mark in marks:
            if mark not in seen:
                seen.add(mark)
                deduplicated.append(mark)

        chain_bins: dict[bytes, list[ProvenanceMark]] = {}
        for mark in deduplicated:
            chain_bins.setdefault(mark.chain_id(), []).append(mark)

        chains: list[ChainReport] = []
        for chain_id, chain_marks in chain_bins.items():
            chain_marks.sort(key=lambda mark: mark.seq())
            first = chain_marks[0]
            has_genesis = first.seq() == 0 and first.is_genesis()
            chains.append(
                ChainReport(
                    chain_id=chain_id,
                    has_genesis=has_genesis,
                    marks=tuple(chain_marks),
                    sequences=tuple(_build_sequence_bins(chain_marks)),
                )
            )

        chains.sort(key=lambda chain: chain.chain_id)
        return ValidationReport(tuple(deduplicated), tuple(chains))

    def format(self, report_format: ValidationReportFormat) -> str:
        if report_format is ValidationReportFormat.Text:
            return self._format_text()
        json_obj = self.to_json_obj()
        if report_format is ValidationReportFormat.JsonCompact:
            return json.dumps(json_obj, separators=(",", ":"), ensure_ascii=False)
        return json.dumps(json_obj, indent=2, ensure_ascii=False)

    def _format_text(self) -> str:
        if not self._is_interesting():
            return ""

        lines: list[str] = [
            f"Total marks: {len(self.marks)}",
            f"Chains: {len(self.chains)}",
            "",
        ]
        for chain_index, chain in enumerate(self.chains, start=1):
            short_chain_id = chain.chain_id_hex()[:8]
            lines.append(f"Chain {chain_index}: {short_chain_id}")
            if not chain.has_genesis:
                lines.append("  Warning: No genesis mark found")
            for sequence in chain.sequences:
                for flagged_mark in sequence.marks:
                    mark = flagged_mark.mark
                    annotations: list[str] = []
                    if mark.is_genesis():
                        annotations.append("genesis mark")
                    for issue in flagged_mark.issues:
                        if issue.issue_type == "SequenceGap":
                            assert issue.data is not None
                            annotations.append(f"gap: {issue.data['expected']} missing")
                        elif issue.issue_type == "DateOrdering":
                            assert issue.data is not None
                            annotations.append(
                                f"date {issue.data['previous']} < {issue.data['next']}"
                            )
                        elif issue.issue_type == "HashMismatch":
                            annotations.append("hash mismatch")
                        elif issue.issue_type == "KeyMismatch":
                            annotations.append("key mismatch")
                        elif issue.issue_type == "NonGenesisAtZero":
                            annotations.append("non-genesis at seq 0")
                        elif issue.issue_type == "InvalidGenesisKey":
                            annotations.append("invalid genesis key")
                    if annotations:
                        lines.append(
                            f"  {mark.seq()}: {mark.id_hex()[:8]} ({', '.join(annotations)})"
                        )
                    else:
                        lines.append(f"  {mark.seq()}: {mark.id_hex()[:8]}")
            lines.append("")
        return "\n".join(lines).rstrip()

    def _is_interesting(self) -> bool:
        if not self.chains:
            return False
        if any(not chain.has_genesis for chain in self.chains):
            return True
        if len(self.chains) == 1:
            chain = self.chains[0]
            if len(chain.sequences) == 1 and all(
                not flagged.issues for flagged in chain.sequences[0].marks
            ):
                return False
        return True

    def has_issues(self) -> bool:
        if any(not chain.has_genesis for chain in self.chains):
            return True
        for chain in self.chains:
            for sequence in chain.sequences:
                if any(flagged.issues for flagged in sequence.marks):
                    return True
        if len(self.chains) > 1:
            return True
        return len(self.chains) == 1 and len(self.chains[0].sequences) > 1

    def to_json_obj(self) -> dict[str, object]:
        return {
            "marks": [mark.ur_string() for mark in self.marks],
            "chains": [chain.to_json_obj() for chain in self.chains],
        }


def _sequence_gap(expected: int, actual: int) -> ValidationIssue:
    return ValidationIssue("SequenceGap", {"expected": expected, "actual": actual})


def _hash_mismatch(expected: bytes, actual: bytes) -> ValidationIssue:
    return ValidationIssue("HashMismatch", {"expected": expected, "actual": actual})


def _date_ordering(previous: Date, next_: Date) -> ValidationIssue:
    return ValidationIssue("DateOrdering", {"previous": previous, "next": next_})


def _build_sequence_bins(marks: list[ProvenanceMark]) -> list[SequenceReport]:
    sequences: list[SequenceReport] = []
    current: list[FlaggedMark] = []
    for index, mark in enumerate(marks):
        if index == 0:
            current.append(FlaggedMark.new(mark))
            continue
        previous = marks[index - 1]
        try:
            previous.precedes_opt(mark)
            current.append(FlaggedMark.new(mark))
        except Error as exc:
            if current:
                sequences.append(_create_sequence_report(current))
            issue = (
                exc.validation_issue
                if isinstance(exc.validation_issue, ValidationIssue)
                else ValidationIssue("KeyMismatch")
            )
            current = [FlaggedMark.with_issue(mark, issue)]
    if current:
        sequences.append(_create_sequence_report(current))
    return sequences


def _create_sequence_report(marks: list[FlaggedMark]) -> SequenceReport:
    start_seq = marks[0].mark.seq() if marks else 0
    end_seq = marks[-1].mark.seq() if marks else 0
    return SequenceReport(start_seq, end_seq, tuple(marks))

